File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        stack<int> st;
        ListNode* current = head;
        while(current != NULL)
        {
            st.push(current->val);
            current = current->next;
        }

        ListNode* p = head;
        int maxSum = 0;
        while(p != NULL)
        {
            int sum = p->val + st.top();
            st.pop();
            maxSum = max(sum, maxSum);
            p = p->next;
        }
        return maxSum;
    }
};

File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public int PairSum(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;

        while(fast.next != null && fast.next.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode temp = slow.next;
        slow.next = null;
        ListNode prev = null;
        ListNode next = null;
        while(temp != null)
        {
            next = temp.next;
            temp.next = prev;
            prev = temp;
            temp = next;
        }

        ListNode cur = head;
        int res = 0;
        while(cur != null)
        {
            res = Math.Max(cur.val + prev.val, res);
            cur = cur.next;
            prev = prev.next;
        }

        return res;
    }
}